File: 42.rb
# https://leetcode.com/problems/trapping-rain-water/

def trap(height)
	lindex = 0 
	rindex = 1
	water = 0
	return water if height.empty?

	height.each_with_index { |currh, i|
		# puts "left: #{lindex}, right: #{rindex}, i: #{i}, water: #{water}"
		if currh < height[lindex]
			rindex = [i+1, height.size-1].min if rindex <= i
			if height[rindex] >= height[lindex]
				water += height[lindex] - currh
			else
				rindex += height[rindex..-1].index(height[rindex..-1].max)
				# puts "####left: #{lindex}, right: #{rindex}, i: #{i}, water: #{water}"
				if height[rindex] > currh
					water += [height[lindex], height[rindex]].min - currh
				end
			end
		else
			lindex = i
		end
	} 
	water
end

File: js/42.js
// https://leetcode.com/problems/trapping-rain-water/
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
	if (!height || height.length < 3) {
		return 0
	}
	let rightMax = [-1]
	let right = height.length-2
	while (right >= 0) {
		rightMax.unshift(Math.max(height[right+1], rightMax[0]))
		right -= 1
	}
	console.log(rightMax.length)
	let leftMax = 0
	let water = 0
	for (let i=1; i<height.length; i++) {
		leftMax = Math.max(leftMax, height[i-1])
		let res = Math.min(leftMax, rightMax[i]) - height[i]
		if (res > 0) {
			water += res
		}
	}
	return water
};
